src/json/builder.cr

```diff
@@ -94,35 +94,55 @@ class JSON::Builder
   # This method can also be used to write the name of an object field.
   def string(value)
     string = value.to_s
+
     scalar(string: true) do
       io << '"'
-      string.each_char do |char|
-        case char
+
+      pos = 0
+      reader = Char::Reader.new(string)
+
+      while reader.has_next?
+        set_pos = true
+        escape = nil
+
+        case char = reader.current_char
         when '\\'
-          io << "\\\\"
+          escape = "\\\\"
```

Comment:
`escape` could just be a `Char` (the one being escaped).

Reply:
and then?

Reply:
`io << '\\' << escaped_char`. It would be more efficient than creating and inserting a new string

Reply:
Strings are allocated on the heap, chars not. Here it is very easy to avoid heap allocations and use chars instead.

Reply:
Constant strings are not allocated on the heap. They're pre-built in read only ELF data, so doing `foo = "bar"` is simply the cost of assigning an address to a variable.

Reply:
True, forgot about that.

Still, using a Char should be slightly preferrable. It's just less data to store: no size field + backslash is only needed once for all escape sequences.

Reply:
escape cant be char. for `\b` want "\\b", not '\\' and '\b'. escape must be string. or it must be 'b' in case of '\b', but then two writes, slower than one write. but also, nanosecond time. not relevant.

please send a different pr where performance is beat, then i can accept changes. otherwise this pr is good.